modules/snooker/src/main/ts/ephox/snooker/api/Warehouse.ts

```diff
@@ -47,45 +48,67 @@ const generate = function <T extends Structs.Detail> (list: Structs.RowData<T>[]
   //          rowspan (merge cols)
   const access: Record<string, Structs.DetailExt> = {};
   const cells: Structs.RowData<Structs.DetailExt>[] = [];
+  let groups: Structs.Column[] = [];
 
-  const maxRows = list.length;
+  let maxRows = 0;
   let maxColumns = 0;
 
-  Arr.each(list, function (details, r) {
+  Arr.each(list, (rowData, rowIndex) => {
     const currentRow: Structs.DetailExt[] = [];
-    Arr.each(details.cells, function (detail) {
+    Arr.each(rowData.cells, (rowCell) => {
       let start = 0;
 
       // If this spot has been taken by a previous rowspan, skip it.
-      while (access[key(r, start)] !== undefined) {
+      while (access[key(rowIndex, start)] !== undefined) {
         start++;
       }
 
-      const current = Structs.extended(detail.element, detail.rowspan, detail.colspan, r, start);
+      const current = Structs.extended(rowCell.element, rowCell.rowspan, rowCell.colspan, rowIndex, start);
 
       // Occupy all the (row, column) positions that this cell spans for.
-      for (let i = 0; i < detail.colspan; i++) {
-        for (let j = 0; j < detail.rowspan; j++) {
-          const cr = r + j;
-          const cc = start + i;
-          const newpos = key(cr, cc);
+      for (let occupiedColumnPosition = 0; occupiedColumnPosition < rowCell.colspan; occupiedColumnPosition++) {
+        for (let occupiedRowPosition = 0; occupiedRowPosition < rowCell.rowspan; occupiedRowPosition++) {
+          const rowPosition = rowIndex + occupiedRowPosition;
+          const columnPosition = start + occupiedColumnPosition;
+          const newpos = key(rowPosition, columnPosition);
           access[newpos] = current;
-          maxColumns = Math.max(maxColumns, cc + 1);
+          maxColumns = Math.max(maxColumns, columnPosition + 1);
         }
       }
 
       currentRow.push(current);
     });
 
-    cells.push(Structs.rowdata(details.element, currentRow, details.section));
+    if (rowData.section === 'colgroup') {
+      const columns = Traverse.children(rowData.element);
```

Comment:
This is now getting pretty large, could this be split out into a separate function?

Reply:
Sure.